src/annotator/integrations/pdf.js

```diff
@@ -309,14 +334,78 @@ export class PDFIntegration {
   }
 
   /**
+   * Scroll to the location of an anchor in the PDF.
+   *
+   * If the anchor refers to a location that is an un-rendered page far from
+   * the viewport, then scrolling happens in three phases. First the document
+   * scrolls to the approximate location indicated by the placeholder anchor,
+   * then `scrollToAnchor` waits until the page's text layer is rendered and
+   * the annotation is re-anchored in the fully rendered page. Then it scrolls
+   * again to the final location.
+   *
    * @param {Anchor} anchor
    */
-  scrollToAnchor(anchor) {
-    // TODO - Account for pages changing state between un-rendered and rendered
-    // while scrolling.
-    const highlights = /** @type {Element[]} */ (anchor.highlights);
-    return new Promise(resolve => {
-      scrollIntoView(highlights[0], resolve);
-    });
+  async scrollToAnchor(anchor) {
+    const annotation = anchor.annotation;
+    const isPlaceholder = anchorIsPlaceholder(anchor);
+    const offset = this._anchorOffset(anchor);
+    if (offset === null) {
+      return;
+    }
+
+    await scrollElement(this.contentContainer(), offset);
+
+    if (isPlaceholder) {
+      const anchor = await this._waitForAnnotationToBeAnchored(
+        annotation,
+        this._reanchoringWait
+      );
+      if (!anchor) {
+        return;
+      }
+      const offset = this._anchorOffset(anchor);
+      if (offset === null) {
+        return;
+      }
+      await scrollElement(this.contentContainer(), offset);
+    }
+  }
+
+  /**
+   * Wait for an annotation to be anchored in a rendered page.
+   *
+   * @param {AnnotationData} annotation
+   * @param {number} maxWait
+   * @return {Promise<Anchor|null>}
+   */
+  async _waitForAnnotationToBeAnchored(annotation, maxWait) {
+    const start = Date.now();
+    let anchor;
+    do {
+      // nb. Re-anchoring might result in a different anchor object for the
```

Comment:
What's the impact if that happens?

_Update_: I'm working through this by following a test case. It believe it means that you may well get a different `anchor` object result when asking the same question (`this.annotator.anchors.find(an anchor for this annotation)`) on different loop iterations. Correct?

Reply:
Yes. Most importantly, the anchor associated with an annotation's tag in `this.annotator.anchor` will be different when the anchor is in a placeholder vs when it is in the real text layer.